device/ArithmeticOperations.cpp

```diff
@@ -27,52 +40,115 @@ PackedFloat Multiply(PackedFloat const &a, PackedFloat const &b) {
-    const bool m_sign = a.sign != b.sign;
-    return {m_sign, m_exponent, &m_mantissa};
+    PackedFloat result;
+    std::memcpy(result.mantissa, &m_mantissa, kMantissaBytes);
+    result.exponent = m_exponent;
+    result.sign = a.sign != b.sign;
+    return result;
 }
 
+// Does this correctly output the result if a and b are different signs?
+// The mantissa of the result should depend on the sign bits of a and b
+
 PackedFloat Add(PackedFloat const &a, PackedFloat const &b) {
 #pragma HLS INLINE
 
     // Figure out how much we need to shift by
     ap_uint<kMantissaBits> a_mantissa(*reinterpret_cast<ap_uint<kMantissaBits> const *>(a.mantissa));
     ap_uint<kMantissaBits> b_mantissa(*reinterpret_cast<ap_uint<kMantissaBits> const *>(b.mantissa));
-    const bool a_is_larger = a.exponent > b.exponent;
-    const bool b_is_larger = !a_is_larger;
-    const bool a_is_zero = a_mantissa == 0;
-    const bool b_is_zero = b_mantissa == 0;
-    const Exponent res_exponent = ((a_is_larger && !a_is_zero) || b_is_zero) ? a.exponent : b.exponent;
-    const Exponent shift_c = (b_is_larger && !b_is_zero) ? (b.exponent - a.exponent) : 0;
-    const Exponent shift_m = (a_is_larger && !a_is_zero) ? (a.exponent - b.exponent) : 0;
-    // Optionally shift by 1, 2, 4, 8, 16... log2(B), such that all bits have eventually traveled to
+
+#ifndef HLSLIB_SYNTHESIS
+    // We better not be getting subnormal inputs
+    assert(a.IsZero() || IsMostSignificantBitSet(a_mantissa));
+    assert(a.IsZero() || IsMostSignificantBitSet(b_mantissa));
+#endif
+
+    const bool exp_are_equal = (a.exponent == b.exponent);
+    const bool a_exp_is_larger = (a.exponent > b.exponent);
+    const bool a_mant_is_zero = a_mantissa == 0;
+    const bool b_mant_is_zero = b_mantissa == 0;
+    const bool a_is_larger = (!a_mant_is_zero) && (a_exp_is_larger || (exp_are_equal && a.mantissa > b.mantissa));
+    const bool subtraction = a.sign != b.sign;
+    Exponent res_exponent = ((a_exp_is_larger && !a_mant_is_zero) || b_mant_is_zero) ? a.exponent : b.exponent;
+    const Exponent shift_c = (!a_exp_is_larger && !b_mant_is_zero) ? (b.exponent - a.exponent) : 0;
+    const Exponent shift_m = (a_exp_is_larger && !a_mant_is_zero) ? (a.exponent - b.exponent) : 0;
+
+#ifndef HLSLIB_SYNTHESIS
+    // Turns out Xilinx allows signed shifts!
+    assert(shift_m >= 0);
+    assert(shift_c >= 0);
+#endif
+
+    // Optionally shift right by 1, 2, 4, 8, 16... log2(B), such that all bits have eventually traveled to
     // their designated position.
-    const int kNumStages = hlslib::ConstLog2(kBits);
-ShiftStages:
-    for (int i = 0; i < kNumStages; ++i) {
-#pragma HLS UNROLL
-        a_mantissa = ((shift_c & (1 << i)) == 0) ? a_mantissa : (a_mantissa >> (1 << i));
-        b_mantissa = ((shift_m & (1 << i)) == 0) ? b_mantissa : (b_mantissa >> (1 << i));
-    }
+    // This is unsigned
+    a_mantissa = a_mantissa >> shift_c;
+    b_mantissa = b_mantissa >> shift_m;
```

Comment:
This is definitely nicer (the old code was a remnant from when we used a language that didn't allow dynamic shifts). Did you compare the reports or the generated RTL between the two, to make sure it does something reasonable?

Reply:
I have not. I was just eliminating code for the moment. If you think this has a reasonable shot at generating sane hardware then I'll give it a shot

Reply:
No clue :-) It might, and it might be bother worse and better than our homemade version, so we have to check